### PR TITLE
[17.06] Add cgroup namespaces 

### PR DESCRIPTION
**Pull request by crosbymichael**
Cherry pick of the upstream change to support cgroup namespaces and consistent process addition to cgroups.

### CODE REVIEW

**Comment by kolyshkin** [comment]
Upstream commit https://github.com/opencontainers/runc/commit/df3fa115f97407713ec6ebe37a5ccf13b81dafd2 was backported with a bug; the hunk changing unshare() in libcontainer/nsenter/nsexec.c was applied incorrectly causing two `unshare()` calls instead of one. Second call fails, leading to EINVAL.

In addition, part of upstream commit https://github.com/opencontainers/runc/commit/e92add215 got mixed in here for some reason. Would be OK if it was complete I guess but it's not.

File: libcontainer/nsenter/nsexec.c [hunk 1]
```diff
@@ -543,8 +549,7 @@ void nsexec(void)
 	 */
 	case JUMP_PARENT: {
 			int len;
-			pid_t child;
-			char buf[JSON_MAX];
+			pid_t child, first_child = -1;
```
Comment (kolyshkin) [inline libcontainer/nsenter/nsexec.c:552]: adding `first_child` here is upstream commit e92add2151e50cd66c; I see some parts of that commit in here. I guess it either needs to be brought in (as a separate commit) or not (in this case `first_child` should not appear here).

File: libcontainer/nsenter/nsexec.c [hunk 2]
```diff
@@ -622,6 +627,18 @@ void nsexec(void)
...
+						 *
+						 * Send the init_func pid and the pid of the first child back to our parent.
+						 * We need to send both back because we can't reap the first child we created (CLONE_PARENT).
+						 * It becomes the responsibility of our parent to reap the first child.
+						 */
+						len = dprintf(pipenum, "{\"pid\": %d, \"pid_first\": %d}\n", child, first_child);
+						if (len < 0) {
+							kill(child, SIGKILL);
+							bail("unable to generate JSON for child pid");
+						}
```
Comment (kolyshkin) [inline libcontainer/nsenter/nsexec.c:641]: ditto (also possibly this ended up in the wrong place)

File: libcontainer/nsenter/nsexec.c [hunk 3]
```diff
@@ -754,6 +759,18 @@ void nsexec(void)
...
+			 * ordering might break in the future (especially with rootless
+			 * containers). But for now, it's not possible to split this into
+			 * CLONE_NEWUSER + [the rest] because of some RHEL SELinux issues.
+			 *
+			 * Note that we don't merge this with clone() because there were
+			 * some old kernel versions where clone(CLONE_PARENT | CLONE_NEWPID)
+			 * was broken, so we'll just do it the long way anyway.
+			 */
+			if (unshare(config.cloneflags & ~CLONE_NEWCGROUP) < 0)
+				bail("failed to unshare namespaces");
```
Comment (kolyshkin) [inline libcontainer/nsenter/nsexec.c:773]: This hunk ended up in the wrong place. It should _replace_ a similar one, but without ` & ~CLONE_NEWCGROUP`.

**Review by kolyshkin — CHANGES_REQUESTED** [review]
need to fix a few things

**Comment by andrewhsu** [comment]
Closing this PR now that #7 is in.

**Pull request closed without merging**